### PR TITLE
Enable act status collect on fln node

### PR DESCRIPTION
**Pull request by KazunoriSumiya**
## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: tests_pytest/common_tests/unit_tests/test_model_collector.py [hunk 1]
```diff
@@ -95,8 +117,7 @@ def test_create_tensor2node_assigns_stats_collector(self, fw_info_mock):
         assigned_node, assigned_collector = args
         assert assigned_node is node
         assert isinstance(assigned_collector, StatsCollector)
-
-
+
```
Comment (gouda-youichi) [inline tests_pytest/common_tests/unit_tests/test_model_collector.py:120]: 不要な変更と思います。改行が一つ消えているのでしょうか。。

File: model_compression_toolkit/core/common/quantization/node_quantization_config.py [hunk 2]
```diff
@@ -206,10 +207,21 @@ def set_activation_quantization_param(self,
             activation_params: Dictionary that contains weight quantization params.
 
         """
-        assert self.quant_mode == ActivationQuantizationMode.QUANT
+        assert self.enable_activation_quantization or self.fln_quantization
```
[the remark this answers is not in the document]
Comment (KazunoriSumiya) [reply model_compression_toolkit/core/common/quantization/node_quantization_config.py]: 変更を元に戻しました。

File: tests_pytest/common_tests/unit_tests/test_model_collector.py [hunk 3]
```diff
@@ -197,9 +272,122 @@ def setup(self, fw_impl_mock, fw_info_mock):
...
+        mc.infer(self.infer_input)
+        # Check that inference runs with gradients enabled.
+        fw_impl_mock.run_model_inference.assert_called_once_with(mc.model, self.infer_input, requires_grad=True)
+        # Confirm that the Hessian data is fetched.
+        mc.hessian_service.fetch_hessian.assert_called_once()
+
+    def test_update_statistics_called(self, fw_impl_mock, fw_info_mock):
+        """
+        Verify that update_statistics is called for each statistics container during inference.
+        """
```
Comment (kkawa14) [inline tests_pytest/common_tests/unit_tests/test_model_collector.py:336]: FLNノードに対してのセットアップをしている(他はFalseとする)旨がない。コメントの追記もしくは修正。

File: model_compression_toolkit/core/common/graph/base_node.py [hunk 4]
```diff
@@ -134,6 +134,12 @@ def is_activation_quantization_enabled(self) -> bool:
         """
         return self._is_single_quant_mode(ActivationQuantizationMode.QUANT)
 
+    def is_fln_quantization(self) -> bool:
+        """
+        Returns: Whether fln node activation quantization is enabled or not.
```
[the remark this answers is not in the document]
Comment (KazunoriSumiya) [reply model_compression_toolkit/core/common/graph/base_node.py]: コメントを
 Whether the node's activation quantization is FLN
に修正しました。

File: tests_pytest/common_tests/unit_tests/test_model_collector.py [hunk 5]
```diff
@@ -178,6 +254,49 @@ def fake_create_tensor2node(graph, node, fw_info):
...
+        assert isinstance(graph.get_out_stats_collector(node3), StatsCollector)
+        assert isinstance(graph.get_in_stats_collector(node2), StatsCollector)
+        assert isinstance(graph.get_in_stats_collector(node3), NoStatsCollector)
+        assert mc.intermediate_output_tensors == [node1]
+        assert mc.model_outputs == [node3]
+        assert len(mc.stats_containers_list) == 2
+
 
 class TestModelCollectorInfer:
     @pytest.fixture(autouse=True)
```
[the remark this answers is not in the document]
Comment (KazunoriSumiya) [reply tests_pytest/common_tests/unit_tests/test_model_collector.py:258]: テスト用クラス
class TestFLNModelCollectorInfer:
を作成しました。

File: tests_pytest/common_tests/unit_tests/test_model_collector.py [hunk 6]
```diff
@@ -197,9 +272,122 @@ def setup(self, fw_impl_mock, fw_info_mock):
...
+        assert_array_equal(calls[0][0][0], self.fake_output1)
+        assert_array_equal(calls[0][0][1], np.abs(self.fake_hessian_result))
+
+        assert_array_equal(calls[1][0][0], self.fake_output2)
+        assert_array_equal(calls[1][0][1], np.abs(self.fake_hessian_result))
+
+        assert calls[2][0][0] is self.fake_output3
+        assert calls[2][0][1] is None
+
+class TestFLNModelCollectorInfer:
```
Comment (kkawa14) [inline tests_pytest/common_tests/unit_tests/test_model_collector.py:356]: test_infer_without_hessian、test_infer_with_hessian、test_update_statistics_calledについて、TestModelCollectorInfer(QUANTモードのみでのテスト)に含まれるメソッド名と同じである。
そのため、flnのテストとわかるようにメソッド名を修正。またコメントについても同様

**Review by kkawa14 — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/common/graph/base_node.py [hunk 7]
```diff
@@ -134,6 +134,12 @@ def is_activation_quantization_enabled(self) -> bool:
         """
         return self._is_single_quant_mode(ActivationQuantizationMode.QUANT)
 
+    def is_fln_quantization(self) -> bool:
+        """
+        Returns: Whether the node's activation quantization is FLN
```
Comment (kawakami-masaki0) [inline model_compression_toolkit/core/common/graph/base_node.py:139]: 練習がてら。。
直訳すると「ノードの活性化量子化がFLNであるかどうか」で、
より適した表現があったら良いなと思いました。

**Review by kawakami-masaki0 — APPROVED** [review]

File: model_compression_toolkit/core/common/quantization/filter_nodes_candidates.py [hunk 8]
```diff
@@ -71,7 +79,7 @@ def _filter_bit_method_dups(candidates: List[CandidateNodeQuantizationConfig],
     return final_candidates
 
 
-def filter_node_candidates(node: BaseNode, fw_info) -> List[CandidateNodeQuantizationConfig]:
+def filter_node_candidates(node: BaseNode, fw_info, op_cfg) -> List[CandidateNodeQuantizationConfig]:
```
Comment (kawakami-masaki0) [inline model_compression_toolkit/core/common/quantization/filter_nodes_candidates.py:82]: op_cfgの型ヒントは不要ですか？

File: tests_pytest/common_tests/unit_tests/test_model_collector.py [hunk 9]
```diff
@@ -178,6 +210,49 @@ def fake_create_tensor2node(graph, node, fw_info):
...
+        assert isinstance(graph.get_out_stats_collector(node3), StatsCollector)
+        assert isinstance(graph.get_in_stats_collector(node2), StatsCollector)
+        assert isinstance(graph.get_in_stats_collector(node3), NoStatsCollector)
+        assert mc.intermediate_output_tensors == [node1]
+        assert mc.model_outputs == [node3]
+        assert len(mc.stats_containers_list) == 2
+
 
 class TestModelCollectorInfer:
     @pytest.fixture(autouse=True)
```
Comment (kkawa14) [inline tests_pytest/common_tests/unit_tests/test_model_collector.py:259]: setupメソッドについて、FLNノードに対してのセットアップをしている(他はFalseとする)旨がない。コメントの追記もしくは修正。

File: tests_pytest/common_tests/unit_tests/test_model_collector.py [hunk 10]
```diff
@@ -197,9 +273,123 @@ def setup(self, fw_impl_mock, fw_info_mock):
...
+        assert_array_equal(calls[1][0][0], self.fake_output2)
+        assert_array_equal(calls[1][0][1], np.abs(self.fake_hessian_result))
+
+        assert calls[2][0][0] is self.fake_output3
+        assert calls[2][0][1] is None
+
+
+class TestFLNModelCollectorInfer:
+    @pytest.fixture(autouse=True)
+    def setup(self, fw_impl_mock, fw_info_mock):
```
Comment (kkawa14) [inline tests_pytest/common_tests/unit_tests/test_model_collector.py:360]: setupメソッドについて、FLNノードに対してのセットアップをしている(他はFalseとする)旨がない。コメントの追記もしくは修正。

**Review by kkawa14 — APPROVED** [review]

**Review by gouda-youichi — APPROVED** [review]
OKです。

**Comment by KazunoriSumiya** [comment]
merge stage 3-3
